### PR TITLE
Make the escape key close windows on the MapWnd in LIFO order.

### PR DESCRIPTION
**Pull request by wheals**
As discussed in http://www.freeorion.org/forum/viewtopic.php?f=28&t=10759.

I tested this on a variety of combinations of windows and it always seemed to work, but there might be edge cases I missed still.

### CODE REVIEW

**Comment by geoffthemedio** [comment]
Default open windows seemingly don't get added to the list, so won't close with escape until closed and reopened. Eg. messages, empires.

**Comment by wheals** [comment]
Good catch, I missed that because my config options don't seem to open any windows by default.

File: UI/MapWnd.cpp [hunk 1]
```diff
@@ -5824,70 +5836,60 @@ void MapWnd::Sanitize() {
...
+        m_wnd_stack.erase(it);
     }
+}
 
-    if (m_production_wnd->Visible()) {
-        ToggleProduction();
-        some_subscreen_was_visible = true;
+bool MapWnd::ReturnToMap() {
+    std::shared_ptr<GG::Wnd> wnd;
+    while (!m_wnd_stack.empty() && !(wnd && wnd->Visible())) {
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: add comment explaining what this is doing and/or why

File: UI/MapWnd.cpp [hunk 2]
```diff
@@ -5824,70 +5836,60 @@ void MapWnd::Sanitize() {
...
-    // close sidepanel if open
-    if (SidePanel::SystemID() != INVALID_OBJECT_ID) {
-       SelectSystem(INVALID_OBJECT_ID);
-       return true;
+    } else if (cui && wnd == cui->GetPlayerListWnd()) {
+        HideEmpires();
+    } else if (cui && wnd == cui->GetMessageWnd()) {
+        HideMessages();
+    } else {
+        ErrorLogger() << "Unknown GG::Wnd " << wnd->Name() << " found in MapWnd::m_wnd_stack";
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp:5900]: are you sure `wnd` is not null here?

File: UI/MapWnd.cpp [hunk 3]
```diff
@@ -5824,70 +5836,60 @@ void MapWnd::Sanitize() {
...
-        ToggleSitRep();
-        some_subscreen_was_visible = true;
-    }
-
-    if (m_research_wnd->Visible()) {
-        ToggleResearch();
-        some_subscreen_was_visible = true;
-    }
+void MapWnd::PushWndStack(std::shared_ptr<GG::Wnd> wnd) {
+    RemoveFromWndStack(wnd);
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: add safety check for null `wnd`

File: UI/MapWnd.cpp [hunk 4]
```diff
@@ -5824,70 +5836,60 @@ void MapWnd::Sanitize() {
...
-
-    // close fleets window if open
-    FleetUIManager& fm = FleetUIManager::GetFleetUIManager();
-    GG::Wnd* active_fleet_wnd = fm.ActiveFleetWnd();
-    if (active_fleet_wnd) {
+    } else if (wnd == m_combat_report_wnd) {
+        m_combat_report_wnd->Hide();
+    } else if (wnd == m_side_panel) {
+        SelectSystem(INVALID_OBJECT_ID);
+    } else if (wnd.get() == fm.ActiveFleetWnd()) {
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp:5893]: also here

File: UI/MapWnd.cpp (hunk 2 above)
Comment (wheals) [reply UI/MapWnd.cpp:5900]: Yes, the check on line 5865 will return from the function if `!wnd`

File: UI/MapWnd.cpp (hunk 4 above)
Comment (wheals) [reply UI/MapWnd.cpp:5893]: Same as the other comment

File: UI/MapWnd.cpp [hunk 5]
```diff
@@ -5837,6 +5837,11 @@ void MapWnd::Sanitize() {
 }
 
 void MapWnd::PushWndStack(std::shared_ptr<GG::Wnd> wnd) {
+    if (!wnd) {
+        return;
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: remove the `{}` or at least put on the same line as `return`

File: UI/MapWnd.cpp [hunk 6]
```diff
@@ -5850,10 +5855,13 @@ void MapWnd::RemoveFromWndStack(std::shared_ptr<GG::Wnd> wnd) {
 
 bool MapWnd::ReturnToMap() {
     std::shared_ptr<GG::Wnd> wnd;
+    // Find the top non-null Wnd in the stack (check if it's visible, in case it was somehow closed
+    // without being removed from the stack).
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: are any such null Wnds removed at some point?

**Pull request merged**